### PR TITLE
Newest version of portal.

### PR DESCRIPTION
**Pull request by teryhill**
Thanks Jerry. This version allows for the patient to self register. I would like to get this merged so I can do a GSoC to modify it into a Kiosk,

### CODE REVIEW

**Comment by aethelwulffe** [comment]
I will get this pulled down for review ASAP.

File: patient_portal/get_appointments.php [hunk 1]
```diff
@@ -54,8 +54,9 @@
         $dispmin  = substr($row['pc_startTime'], 3, 2);
         if ($disphour >= 12) {
             $dispampm = "pm";
```
Comment (aethelwulffe) [inline patient_portal/get_appointments.php]: AM/PM is not a standard in Medical settings, or at all for most countries:  Just former British colonies.

> 
> The 24-hour clock is the convention of time keeping in which the day runs from midnight to midnight and is divided into 24 hours, indicated by the hours passed since midnight, from 0 to 23. This system is the most commonly used time notation in the world today,[1] and is used by international standard ISO 8601.[2]
> 
> A limited number of countries, particularly English-speaking nations, use the 12-hour clock as a standard, or a mixture of the 24- and 12-hour time systems. In countries where the 12-hour clock is still dominant, some professions prefer to use the 24-hour clock. For example, in the practice of medicine the 24-hour clock is generally used in documentation of care as it prevents any ambiguity as to when events occurred in a patient's medical history.[3] In the United States and a handful of other countries, it is popularly referred to as military time.[4][1]

File: assets/js/summernote-0-8-2/dist/plugin/nugget/summernote-ext-nugget.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                ]);
+
+                // create jQuery object from button instance.
+                return button.render();
+            });
+        }
+
+    });
+
+}));
```
Comment (aethelwulffe) [inline assets/js/summernote-0-8-2/dist/plugin/nugget/summernote-ext-nugget.js]: newline

File: patient_portal/get_patient_info.php [hunk 3]
```diff
@@ -103,104 +100,121 @@
...
-          ." WHERE ".COL_POR_USER."=?";
-            $auth = privQuery($sql, array($_POST['uname']));
-            if($auth===false)
-            {
+    $sql = "SELECT " . implode(",", array(
+        COL_ID,
+        COL_PID,
+        COL_POR_PWD,
+        COL_POR_SALT,
+        COL_POR_PWD_STAT
```
Comment (aethelwulffe) [inline patient_portal/get_patient_info.php]: arrays defined in multiple lines should add a trailing comma.  This is the "long array" format, and makes adding new ones easier (without making a change to existing lines.
-No big deal here, as this is probably a permanently enumerated array.

File: patient_portal/get_patient_info.php [hunk 4]
```diff
@@ -103,104 +100,121 @@
...
                 {
                     $logit->portalLog('login attempt','',($_POST['uname'].':pass not salted'),'','0');
                     session_destroy();
-                    header('Location: '.$landingpage.'&w');
-                    exit;                        
+                    header('Location: ' . $landingpage . '&w&p');
+                    exit();
                 }
                 $new_salt=oemr_password_salt();
                 $new_hash=oemr_password_hash($plain_code,$new_salt);
```
Comment (aethelwulffe) [inline patient_portal/get_patient_info.php]: REMINDER:  one-day, we should dump oemr for LHEHR everywhere...

File: patient_portal/get_patient_info.php [hunk 5]
```diff
@@ -103,104 +100,121 @@
...
+if ($userData = sqlQuery($sql, array(
+    $auth['pid']
+))) { // if query gets executed
 
         if (empty($userData)) {
             $logit->portalLog('login attempt','',($_POST['uname'].':not active patient'),'','0');
             session_destroy();
                             header('Location: '.$landingpage.'&w');
-            exit;
+        exit();
```
Comment (aethelwulffe) [inline patient_portal/get_patient_info.php]: Either syntax works here, but for a normal exit, if you want to put parens after, you might as well do exit(0);

File: patient_portal/import_template_ui.php [hunk 6]
```diff
@@ -13,52 +13,59 @@
...
+ * Please help the overall project by sending changes you make to the authors and to the LibreHealth EHR community.
  *
  */
 $sanitize_all_escapes=true;
 $fake_register_globals=false;
 require_once("../interface/globals.php");
 $getdir = isset($_POST['sel_pt']) ? $_POST['sel_pt'] : 0;
 if( $getdir > 0){
-    $tdir = $GLOBALS['OE_SITE_DIR'] .  '/onsite_portal_documents/templates/' . $getdir . '/';
+    $tdir = $GLOBALS['OE_SITE_DIR'] .  '/documents/onsite_portal_documents/templates/' . $getdir . '/';
```
Comment (aethelwulffe) [inline patient_portal/import_template_ui.php]: onsite_ is a redacted usage.  I think that patient_portal_documents is a lot more descriptive.

File: patient_portal/index.php [hunk 7]
```diff
@@ -104,23 +102,26 @@
...
 <head>
     <title><?php echo xlt('Patient Portal Login'); ?></title>
 
-    <script type="text/javascript" src="<?php echo $GLOBALS['standard_js_path']; ?>/jquery-min-1-11-3/index.js"></script>
-    <script type="text/javascript" src="<?php echo $GLOBALS['standard_js_path']; ?>/jquery.gritter-1-7-4/js/jquery.gritter.min.js"></script>
-
-    <link rel="stylesheet" type="text/css" href="<?php echo $GLOBALS['standard_js_path']; ?>/jquery.gritter-1-7-4/css/jquery.gritter.css" />
+    <link rel="stylesheet" href="<?php echo $GLOBALS['standard_js_path']; ?>jquery-datetimepicker-2-5-4/build/jquery.datetimepicker.min.css">
+    <script type="text/javascript" src="<?php echo $GLOBALS['standard_js_path']; ?>jquery-min-1-11-3/index.js"></script>
+    <script type="text/javascript" src="<?php echo $GLOBALS['standard_js_path']; ?>jquery.gritter-1-7-4/js/jquery.gritter.min.js"></script>
```
Comment (aethelwulffe) [inline patient_portal/index.php]: Gritter and all these have include_headers entries that can be used.

File: patient_portal/index.php [hunk 8]
```diff
@@ -194,107 +195,209 @@ function validate_new_pass() {
...
+if (isset($_GET['requestNew'])) {
+    $_SESSION['patient_portal_onsite'] = true;
+    $_SESSION['authUser'] = 'portal-user';
+    $_SESSION['pid'] = true;
+    ?>
+    $('.datepicker').datetimepicker({
+        <?php $datetimepicker_timepicker = false; ?>
+        <?php $datetimepicker_showseconds = false; ?>
+        <?php $datetimepicker_formatInput = false; ?>
+        <?php require($GLOBALS['srcdir'] . '/js/xl/jquery-datetimepicker-2-5-4.js.php'); ?>
```
Comment (aethelwulffe) [inline patient_portal/index.php]: Does this exist?

File: patient_portal/lib/appsql.class.php [hunk 9]
```diff
@@ -260,11 +273,18 @@ public function escapeHtml( $string ){
      * @param Date format set in GLOBALS
      * @return Date format in PHP
      */
-    public function dateFormat( $format ){
-        if( $format == "0" ) $date_format = 'yyyy/mm/dd';
-        else if( $format == 1 ) $date_format = 'mm/dd/yyyy';
-        else if( $format == 2 ) $date_format = 'dd/mm/yyyy';
-        else $date_format = $format;
+    public function dateFormat($format)
```
Comment (aethelwulffe) [inline patient_portal/lib/appsql.class.php]: We have date formats defined in globals.

File: patient_portal/lib/doc_lib.php [hunk 10]
```diff
@@ -52,12 +50,18 @@
 
 try{
     $form_filename = $_REQUEST['docid'] . '_' . $GLOBALS['pid'] . '.pdf';
-    $templatedir = $GLOBALS['OE_SITE_DIR'] . "/../../patient_portal/patient_documents";
+    $templatedir = $GLOBALS['OE_SITE_DIR'] . "/documents/onsite_portal_documents/patient_documents";
```
Comment (aethelwulffe) [inline patient_portal/lib/doc_lib.php]: If these are template directories, the directory should not be named "patient_documents".

File: patient_portal/messaging/messages.php [hunk 11]
```diff
@@ -50,29 +50,32 @@
...
-    $dashuser = getUserIDInfo(IS_DASHBOARD);
+    $dashuser = getUserIDInfo($_SESSION['authUserID']);
 }
-function getAuthPortalUsers(){
+
+function getAuthPortalUsers()
+{
     $resultpd = $resultusers= $resultpatients = array ();
     if( IS_DASHBOARD ){ // admin can mail anyone
-        $resultusers[] = ['userid'=>'99998','username'=>'-- Auth Users --'];
```
Comment (aethelwulffe) [inline patient_portal/messaging/messages.php]: Glad to see this gone...

File: patient_portal/messaging/messages.php [hunk 12]
```diff
@@ -50,29 +50,32 @@
...
             $resultpatients[] = $row;
         }
         $resultpd[] = array_merge($resultusers,$resultpatients);
         return $resultpd[0];
-    }
-    else{
+    } else{
         $resultpd = array ();
-        $authusers = sqlStatement( "SELECT users.id as userid, CONCAT(users.fname,' ',users.lname) as username  FROM users WHERE authorized = 1" );
+        $authusers = sqlStatement("SELECT users.username as userid, CONCAT(users.fname,' ',users.lname) as username  FROM users WHERE authorized = 1");
```
Comment (aethelwulffe) [inline patient_portal/messaging/messages.php]: If using the user name,  should keep it username....just to pick...

File: patient_portal/messaging/messages.php [hunk 13]
```diff
@@ -110,14 +114,14 @@ function getAuthPortalUsers(){
...
-app.controller('inboxCtrl', ['$scope', '$filter','$http', function ($scope, $filter,$http) {
+    app.controller('inboxCtrl', ['$scope', '$filter','$http','$window', function ($scope, $filter,$http,$window) {
     $scope.date = new Date;
     $scope.sortingOrder = 'id';
     $scope.pageSizes = [5,10,20,50,100];
     $scope.reverse = false;
     $scope.filteredItems = [];
     $scope.groupedItems = [];
-        $scope.itemsPerPage = 50;
+    $scope.itemsPerPage = 8;
```
Comment (aethelwulffe) [inline patient_portal/messaging/messages.php]: Is this related to the pageSizes[] above?  "8" is not in that arrary.

File: patient_portal/messaging/messages.php [hunk 14]
```diff
@@ -211,35 +220,40 @@ function getAuthPortalUsers(){
...
+            .success(function(data, status, headers, config) {
+                $window.location.reload();
+            }).error(function(data, status, headers, config) {
+                alert(data);
+            });
+            return false;
         };
 
-        $scope.deleteMessage = function (id){ /* @todo add array for mass delete */
+        $scope.deleteMessage = function (id){
```
Comment (aethelwulffe) [inline patient_portal/messaging/messages.php]: Was a mass delete implemented?  Sounds like a darn useful feature...

File: patient_portal/messaging/messages.php [hunk 15]
```diff
@@ -556,127 +630,207 @@ class="pull-right glyphicon glyphicon-warning-sign text-danger"></span></span> -
...
+                <!-- /.modal-dialog -->
+            </div>
+            <!-- /.modal compose message -->
+        </div>
+        <!--/row ng-controller-->
+    </div>
+    <!--/container--> </ng>
 
 </body>
 </html>
```
Comment (aethelwulffe) [inline patient_portal/messaging/messages.php]: newline

File: sql/1_0_1-to-2_0_0_upgrade.sql [hunk 16]
```diff
@@ -210,135 +210,6 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
   ALTER TABLE clinical_rules ADD COLUMN pqrs_code varchar(35) DEFAULT NULL COMMENT 'Measure number';
 #EndIf
 
-#IfNotTable onsite_mail
```
Comment (aethelwulffe) [inline sql/1_0_1-to-2_0_0_upgrade.sql]: Tables not required anymore?

Comment (teryhill) [reply sql/1_0_1-to-2_0_0_upgrade.sql]: They were in the file twice

File: patient_portal/messaging/messages.php (hunk 14 above)
Comment (teryhill) [reply patient_portal/messaging/messages.php]: Yes you can select messages and delete all

File: patient_portal/messaging/messages.php (hunk 12 above)
Comment (teryhill) [reply patient_portal/messaging/messages.php]: Will make that suggestion

File: patient_portal/lib/doc_lib.php (hunk 10 above)
Comment (teryhill) [reply patient_portal/lib/doc_lib.php]: They are template and the current naming difference will be needed with the Kiosk portion of the application

File: patient_portal/lib/appsql.class.php (hunk 9 above)
Comment (teryhill) [reply patient_portal/lib/appsql.class.php]: This is old code that was reused for the code base.

File: patient_portal/index.php (hunk 8 above)
Comment (teryhill) [reply patient_portal/index.php]: Yes

File: patient_portal/index.php (hunk 7 above)
Comment (teryhill) [reply patient_portal/index.php]: Will apply the header code in subsequent releases

File: patient_portal/import_template_ui.php (hunk 6 above)
Comment (teryhill) [reply patient_portal/import_template_ui.php]: Kiosk will have a directory here also

File: patient_portal/get_appointments.php (hunk 1 above)
Comment (teryhill) [reply patient_portal/get_appointments.php]: This is old code that was reused. The add_edit_event has the same items

File: patient_portal/find_appt_popup_user.php [hunk 17]
```diff
@@ -65,24 +64,35 @@
...
+    }
+
   for (; $i < $iend; ++$i) {
    if ($catid == 2) {        // in office
     // If a category ID was specified when this popup was invoked, then select
     // only IN events with a matching preferred category or with no preferred
     // category; other IN events are to be treated as OUT events.
     if ($input_catid) {
-     if ($prefcatid == $input_catid || !$prefcatid)
+                if ($prefcatid == $input_catid || !$prefcatid) {
```
[the remark this answers is not in the document]
Comment (teryhill) [reply patient_portal/find_appt_popup_user.php]: Correct but in the current system the first (I think) eleven are predefined and should not be changed. Also this is old code that was reused and the changes will be made to the rewrite of all the code.

File: patient_portal/add_edit_event_user.php [hunk 18]
```diff
@@ -338,7 +347,9 @@
  $startampm = '1';
  if ($starttimeh >= 12) { // p.m. starts at noon and not 12:01
   $startampm = '2';
-  if ($starttimeh > 12) $starttimeh -= 12;
+    if ($starttimeh > 12) {
```
[the remark this answers is not in the document]
Comment (teryhill) [reply patient_portal/add_edit_event_user.php]: add_edit_event code was used here the changes will need to be made in all the code when re-written

File: patient_portal/find_appt_popup_user.php [hunk 19]
```diff
@@ -102,18 +112,24 @@ function doOneDay($catid, $udate, $starttime, $duration, $prefcatid) {
...
- if ($_REQUEST['searchdays']) $searchdays = $_REQUEST['searchdays'];
+if ($_REQUEST['searchdays']) {
+    $searchdays = $_REQUEST['searchdays'];
+}
 
  // Get a start date.
- if ($_REQUEST['startdate'] && preg_match("/(\d\d\d\d)\D*(\d\d)\D*(\d\d)/",
-     $_REQUEST['startdate'], $matches))
- {
+if ($_REQUEST['startdate'] && preg_match(
```
[the remark this answers is not in the document]
Comment (teryhill) [reply patient_portal/find_appt_popup_user.php]: This is for displaying the selection list. This is old code that is being reused here and when it is changed it will be done system wide

File: patient_portal/messaging/messages.php (hunk 15 above)
Comment (teryhill) [reply patient_portal/messaging/messages.php]: done

File: ccdaservice/ccda_gateway.php [hunk 20]
```diff
@@ -0,0 +1,122 @@
+<?php
```
[the remark this answers is not in the document]
Comment (teryhill) [reply ccdaservice/ccda_gateway.php]: The CCDA Code is a work in process right now. There are a number of changes that are being worked on.

**Comment by aethelwulffe** [comment]
This is a big stretch PR in whatever case.  When we merge, we should put in an issue to attend to some of the date and asset directory type items above.  We need someone else to review this please.

File: assets/js/summernote-0-8-2/dist/plugin/nugget/summernote-ext-nugget.js (hunk 2 above)
Comment (teryhill) [reply assets/js/summernote-0-8-2/dist/plugin/nugget/summernote-ext-nugget.js]: done
